api/v1alpha4/clusterclass_types.go

```diff
@@ -0,0 +1,108 @@
+// +kubebuilder:resource:path=clusterclasses,shortName=cc,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+
+// ClusterClass is a template which can be used to create managed topologies.
+type ClusterClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec ClusterClassSpec `json:"spec,omitempty"`
+}
+
+// ClusterClassSpec describes the desired state of the ClusterClass.
+type ClusterClassSpec struct {
+	// ControlPlane is a reference to a local struct that holds the details
+	// for provisioning the Control Plane for the Cluster.
+	ControlPlane LocalObjectTemplate `json:"controlPlane,omitempty"`
+
+	// Workers describes the worker nodes for the cluster.
+	// It is a collection of node types which can be used to create
+	// the worker nodes of the cluster.
+	// +optional
+	Workers WorkersClass `json:"workers,omitempty"`
+
+	// Infrastructure is a reference to a provider-specific template that holds
+	// the details for provisioning infrastructure specific cluster
+	// for the underlying provider.
+	// The underlying provider is responsible for the implementation
+	// of the template to an infrastructure cluster.
+	Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// WorkersClass is a collection of deployment classes.
+type WorkersClass struct {
+	// MachineDeployments is a list of machine deployment classes that can be used to create
+	// a set of worker nodes.
+	MachineDeployments []MachineDeploymentClass `json:"machineDeployments,omitempty"`
+}
+
+// MachineDeploymentClass serves as a template to define a set of worker nodes of the cluster
+// provisioned using the `ClusterClass`.
+type MachineDeploymentClass struct {
+	// Class denotes a type of worker node present in the cluster,
+	// this name MUST be unique for this Class and can be referenced
+	// in the Cluster to create a managed MachineDeployment.
+	Class string `json:"class"`
+
+	// Template is a local struct containing a collection of templates for creation of
+	// MachineDeployment objects representing a set of worker nodes.
+	Template MachineDeploymentClassTemplate `json:"template"`
+}
+
+// MachineDeploymentClassTemplate defines how a Deployment generated from a MachineDeploymentClass
+// should look like.
+type MachineDeploymentClassTemplate struct {
+	Metadata ObjectMeta `json:"metadata,omitempty"`
+
+	// Bootstrap contains the bootstrap template reference to be used
+	// for the creation of worker Machines.
+	Bootstrap LocalObjectTemplate `json:"bootstrap,omitempty"`
+
+	// Infrastructure contains the infrastructure template reference to be used
+	// for the creation of worker Machines.
+	Infrastructure LocalObjectTemplate `json:"infrastructure,omitempty"`
+}
+
+// LocalObjectTemplate defines a template for a topology Class.
+type LocalObjectTemplate struct {
+	// Ref is a required reference to a custom resource
+	// offered by a bootstrap provider.
```

Comment:
```suggestion
```
Probably best to just drop it. Or alternatively replace it through something like `// offered by a bootstrap, controlplane or infrastructure provider.`